### PR TITLE
gh-89739: gh-77140: Support zip64 in zipimport

### LINKED ISSUE #89739: Cannot import modules from Zip64 files
BPO | [45576](https://bugs.python.org/issue45576)
--- | :---
Nosy | @luisffranca

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-10-22.16:40:33.579>
labels = ['type-bug', 'library', '3.9']
title = 'Cannot import modules from Zip64 files'
updated_at = <Date 2021-10-22.16:40:33.579>
user = 'https://github.com/luisffranca'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-22.16:40:33.579>
actor = 'lfamorim'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-10-22.16:40:33.579>
creator = 'lfamorim'
dependencies = []
files = []
hgrepos = []
issue_num = 45576
keywords = []
message_count = 1.0
messages = ['404792']
nosy_count = 1.0
nosy_names = ['lfamorim']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue45576'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-94146

### PR DESCRIPTION
**Pull request by thatch**
* Reads zip64 files as produced by the zipfile module
* Include tests (somewhat slow, however, because of the need to create "large" zips)
* About the same amount of strictness reading invalid zip files as zipfile has
* Still works on files with prepended data (like pex)

There are a lot more test cases at https://github.com/thatch/zipimport64/ that give me confidence that this works for real-world files.  I assume this relnotes worthy, let me know any other followups you'd like.

As an aside, in its current state can be applied as a patch at least as far back as 3.8, as long as you regenerate frozen modules.  I don't expect cpython to do this, but I certainly will on our internal version.

Fixes #89739 and #77140

* Issue: gh-89739

* Issue: gh-77140

### CODE REVIEW

**Comment by thatch** [comment]
One question I've got is whether the only exception zipimport is allowed to raise is `ZipImportError` or if that's just leftover from the C implementation.  The calls to `.pop(0)` can raise `IndexError` but there may be other non-`ZipImportError` and I haven't fuzzed it.

**Comment by brettcannon** [comment]
`zipimport` can raise whatever exceptions makes sense. `ZipImportError` is specific to custom exceptions that `zipimport` needs to raise.

**Comment by vstinner** [comment]
> @[lisroach](https://github.com/lisroach) lisroach requested a review from [vstinner](https://github.com/vstinner) 22 days ago

I'm not sure that I good candidate for that. @serhiy-storchaka rewrote zipimport in pure Python if I recall correctly. @serhiy-storchaka: would you mind to review this PR?

**Comment by thatch** [comment]
Hi @brettcannon @vstinner 

I've tried to to at-mention @serhiy-storchaka as well as emailed privately a month ago, and have gotten no response.  I know that he's active, e.g. https://github.com/python/cpython/commit/75a6fadf369315b27e12f670e6295cf2c2cf7d7e from yesterday but I'm at a loss for how to actually get his attention.

I have a couple more zip-related fixes I'd like to get in for 3.12 as well (https://github.com/python/cpython/issues/63281 being the one I've hit for work with inplace builds), so would like to discuss an alternative reviewer to get this unblocked.  This PR intentionally makes minimal changes in the non-zip64 path of `_read_directory`, to not reject any zips that currently work.

Potential people that have dealt with large zipimport include @gpshead and @Yhg1s at Google (who I assume patched this internally and never upstreamed).  The "[ask Thomas](https://github.com/python/cpython/issues/75074#issuecomment-1093753448)" reference appears to also recognize him as an expert.

**Comment by gpshead** [comment]
I'll get this reviewed for 3.12.  I like the hack you used in the tests to force it to generate a zip64 format central directory.  i wonder if we also need a test that stores a huge file, but I've barely skimmed the code and will decide that later.

---

Side note: Please don't assume any specific committer can dedicate their volunteer interrupt time to your needs and that some people have more important priorities; for example https://twitter.com/SerhiyStorchaka/status/1579376282252742656 (and more recent sad news).

**Review by vstinner — COMMENTED** [review]
Would it make sense to add a ``_zipfile`` module to share code between zipfile and zipimport? I expect that some code is in common, no?

**Comment by gpshead** [comment]
> Would it make sense to add a `_zipfile` module to share code between zipfile and zipimport? I expect that some code is in common, no?

That kind of refactoring deserves its own issue, it isn't in scope here. You can construct an interpreter with a zipped up standard library if you add zipimport to your bootstrap. Adding dependencies gets complicated. Otherwise zipimport would just use zipfile directly already. See the https://github.com/python/cpython/issues/69897 discussion when the old C zipimport was rewritten into this Python one.

**Comment by vstinner** [comment]
> That kind of refactoring deserves its own issue, it isn't in scope here.

I expect that the review should be easier if changes specific to zipimport are smaller, if more code is shared between zipimport and zipfile.

>  Adding dependencies gets complicated.

Why is it complicated? _zipimport (common code) should be designed to be usable at Python startup, by zipimport. The code which doesn't fit with Python startup should stay in zipfile.

For example, I see that many constants are the same between the two files. Can't they be shared?

```
END_CENTRAL_DIR_SIZE = 22
END_CENTRAL_DIR_SIZE_64 = 56
END_CENTRAL_DIR_LOCATOR_SIZE_64 = 20
STRING_END_ARCHIVE = b'PK\x05\x06'  # standard EOCD signature
STRING_END_LOCATOR_64 = b'PK\x06\x07'  # Zip64 EOCD Locator signature
STRING_END_ZIP_64 = b'PK\x06\x06'  # Zip64 EOCD signature
MAX_COMMENT_LEN = (1 << 16) - 1
MAX_UINT32 = 0xffffffff
ZIP64_EXTRA_TAG = 0x1
```

Well, I'm just talking for myself. If someone else is available to review such large change, please go ahead with the proposed design ;-)

**Comment by thatch** [comment]
>> Adding dependencies gets complicated.
> Why is it complicated? _zipimport (common code) should be designed to be usable at Python startup, by zipimport. The code which doesn't fit with Python startup should stay in zipfile.

It's adding one more frozen module, but those are cheap and plentiful these days.  The constants are one thing, but the logic would be even better to unify.  For example, zipimport doesn't verify crc (which makes https://github.com/python/cpython/issues/63281 easier to hit).  That's a bigger project that I can look at for 3.13, but there are almost certainly behavior changes involved (around which one is chosen for duplicated member names, for one).

File: Lib/test/test_zipimport.py
Comment (brettcannon) [inline Lib/test/test_zipimport.py]:
```suggestion
        # This is the simplest way to make zipfile generate the zip64 EOCD block.
```

Should this be broken out into a utility function? Or have a separate `TestCase` for ZIP64 where this is part of the `setUp()` method?

**Comment by gpshead** [comment]
This needs updating and zipimport merge conflicts resolved.  Given 3.12beta1 happens Monday lets assume this is already missing 3.12 and targeting 3.13 so the versionchanged also needs updating.

That also gives us more time to reconsider if we should do some refactoring to share anything common a new _zipimport frozen module as @vstinner was suggesting.

**Comment by itamaro** [comment]
@gpshead I merged main and updated the PR - buildbots look happy!

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: compute the maximum data that we'd need to read before computing max_comment_start.

```
max_comment_plus_dirs_size = ... # the - pile - of - constants - above
max_comment_start = max(file_size - max_comment_plus_dirs_size, 0)
```

and limit the read as such: `fp.read(max_comment_plus_dirs_size)`

Unbounded read() calls have a bad smell (I realize this came from the previous version).

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: Align this form of hanging indent after the `(` above, not on the `(`.   Fix the same for all other edits involving ` path=archive)` second lines on `ZipImportError` raises in this PR.

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: Be specific about the detail that was wrong in the error message. `f"Expected {END_CENTRAL_DIR_SIZE_64} byte zip64 central directory, but read {len(buffer)} bytes."` or similar.

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: Other code uses _unpack_uint32 and similar.  is there an _unpack_uint64?  Be consistent.  (code using the `_unpack_uint*` functions likely came from before `int.from_bytes` was added to the language in 3.2?)

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: if you add `extra_data = memoryview(extra_data)` here, it'll prevent the slicing in the loop below from creating copies in new bytes objects.

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: Likely faster and capped at a fixed amount of work regardless of input size.

```python
num_extra_values = (len(extra_data) - 4) // 8
if num_extra_values > 3:
   raise ...
values = struct.unpack_from(f"<{min(num_extra_values, 3)}Q", extra_data, offset=4)
```

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: central_directory_position

File: Misc/NEWS.d/next/Library/2022-06-22-14-45-32.gh-issue-89739.CqZcRL.rst
Comment (gpshead) [inline Misc/NEWS.d/next/Library/2022-06-22-14-45-32.gh-issue-89739.CqZcRL.rst]:
```ReST
The :mod:`zipimport` module can now read ...
```

File: Lib/zipimport.py
Comment (gpshead) [inline Lib/zipimport.py]: Don't talk and what's in the test in a comment in the main sources. just end this sentence with "that use all 3 are >4GiB".

relatedly... I don't obviously see what you describe in a test?

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Lib/zipimport.py
Comment (itamaro) [reply Lib/zipimport.py]: there is no `_unpack_uint64` - should I add it in _bootstrap_external?

File: Lib/zipimport.py
Comment (itamaro) [reply Lib/zipimport.py]: > relatedly... I don't obviously see what you describe in a test?

I think this refers to the extra tests in https://github.com/thatch/zipimport64/

**Comment by itamaro** [comment]
I didn't expect the Spanish Inquisition!

**Comment by itamaro** [comment]
> I didn't expect the Spanish Inquisition!

bedevere doesn't trigger on the FUN_TRIGGER_PHRASE anymore?

**Comment by AlexWaygood** [comment]
> > I didn't expect the Spanish Inquisition!
> 
> bedevere doesn't trigger on the FUN_TRIGGER_PHRASE anymore?

I think you have to be the original author of the PR or he's not really interested in anything you have to say 😞

**Comment by itamaro** [comment]
hey @gpshead, other than the `_unpack_uint64` question, this PR should be good to go. PTAL :)

**Comment by gpshead** [comment]
I won't have time to look at this for a several more weeks, but the one thing in the back of my mind about the zipimport implementation is... for import purposes, we should probably be pedantic about the zip format:

Anytime zip allows there to be multiple directory entries or copies of directory information referring to the same file we should ensure that all of them exist as expected and are in agreement.  I haven't looked to see if the current code _already_ does this (central directory vs per file header) - zip64 might provide another opportunity for that kind of shenanigan? _(I have to reread and refresh myself on zip format specifics anytime I dive in on such code)_

I'm thinking of this from a security perspective: we don't want zip files that zipimport's implementation happens to interpret in a different way than any other zip file reader implementation as that could provide ways to potentially hide code.

---

Overall: thanks for everybody's work on this, we'll eventually get it in.

**Comment by itamaro** [comment]
@gpshead I'd like to get this PR past the finish line, but I am not familiar enough with zip format intricacies to follow-up on your last comment, so I'd appreciate it if you could provide additional guidance!
sounds like the best approach might be to add more test cases that cover the security-oriented cases you're concerned about? if you could propose such test cases, or provide pointers that could help me with coming up with such test cases, this might move us forward!

fwiw, I just merged main into this branch locally and it merges cleanly and the test suite still passes, so we should be in a good spot to pick this up.

**Comment by thatch** [comment]
@gpshead As an example, both `zipfile` and `zipimport` are not safe against (for example) zip bombs today.  My goal in writing it this way was to take valid zips that did not work (at all) before, and make them work like they do in zipfile, so if anything I'm aiming for _parity_ as a goal.

I hope that the security-focused aspect, which is quite possibly a breaking change for real-world files _that work today_, can be handled in another issue -- quite likely a PEP, and actually unifying the code for zipfile and zipimport.  Both libraries handle files(that do not conform to what I would consider "zip: the good parts" ([ISO/IEC 21320-1:2015](https://www.iso.org/standard/60101.html) is a start), and I'm almost certain I can construct non-zip64 zip files that are already handled differently between zipfile and zipimport.  There's only a _very_ restrictive subset that is handled universally among decompressors without bugs, and we'd need to have a much larger discussion about what features to prohibit.

Even decompressors that have been around for ages (like Info-Zip) have behavior differences that I would call bugs -- either accepting invalid files or rejecting valid-to-the-appnote ones.  Sorry this is so complicated.

**Comment by thatch** [comment]
> zip64 might provide another opportunity for that kind of shenanigan?

If you're asking if this is one more branch case that decompressors could handle differently, the answer is "yes."  I don't know whether that's a blocker in my mind -- even unifying the two python libs would still have differences when compared with e.g. the "unzip" command on *nix.

**Review by gpshead — APPROVED** [review]
While there's definitely a pile of things I see in zipimport and zipfile module code that bother me... all of that is existing design patterns and should not hold this up. This seems fairly self contained. Has a basic test. & I know you'll be (read: presumably already have been) exercising it a lot more.

Getting this in for more 3.13 alpha and beta baking sounds great.  Thanks for this work!

File: Lib/zipimport.py
Comment (jsirois) [inline Lib/zipimport.py:520]: Where does this symbol come from? I see no struct module import in this file. My Pex too-big-zip tests also find the following under 3.13.0a6:
```
Failed checking if argv[0] is an import path entry
Traceback (most recent call last):
  File "<frozen zipimport>", line 98, in __init__
  File "<frozen zipimport>", line 520, in _read_directory
NameError: name 'struct' is not defined. Did you forget to import 'struct'?
  File "/tmp/pytest-of-jsirois/pytest-10/test_check0/too-big.pyz", line 1
    PK-
SyntaxError: source code cannot contain null bytes
```

Comment (jsirois) [reply Lib/zipimport.py:520]: A quick experiment seems to indicate just adding the import gets things working. I'll try my hand at sending up a patch.

Comment (itamaro) [reply Lib/zipimport.py:520]: good catch! too bad the tests didn't catch it. let me know if you need help with getting that PR going.

Comment (jsirois) [reply Lib/zipimport.py:520]: I think I'm good, thank you. Filed #118107 to reference in the forthcoming PR.

Comment (jsirois) [reply Lib/zipimport.py:520]: Fix here hopefully: #118108

File: Lib/zipimport.py
Comment (jsirois) [inline Lib/zipimport.py:527]: These were problematic as well: `values` is a tuple; not a list. The new test caught this issue.

**Comment by serhiy-storchaka** [comment]
The [zip file related](https://github.com/python/cpython/labels/zip%20file%20related) label is only used for this PR. Was it added by accident? For all other ZIP file related issues we use the [Zipfile issues](https://github.com/orgs/python/projects/7) project.

**Pull request merged**
